x-pack/filebeat/module/zookeeper/audit/ingest/pipeline.yml

```diff
@@ -0,0 +1,103 @@
+description: Pipeline for parsing ZooKeeper audit messages
+processors:
+- set:
+    field: event.ingested
+    value: '{{_ingest.timestamp}}'
+- rename:
+    field: message
+    target_field: log.original
+    ignore_missing: true
+- grok:
+    field: log.original
+    trace_match: true
+    patterns:
+      - '%{TIMESTAMP_ISO8601:zookeeper.audit.timestamp}%{SPACE}%{LOGLEVEL:log.level}%{SPACE}%{CALLER_CLASS:log.logger}:%{SPACE}%{GREEDYDATA:message}'
+      - '%{GREEDYDATA:message}'
+    pattern_definitions:
+      CALLER_CLASS: (%{JAVACLASS}|%{NOTSPACE})
+- rename:
+    field: '@timestamp'
+    target_field: event.created
+    if: ctx?.zookeeper?.audit?.timestamp != null
+- date:
+    if: ctx?.zookeeper?.audit?.timestamp != null && ctx.event.timezone == null
+    field: zookeeper.audit.timestamp
+    target_field: '@timestamp'
+    formats:
+    - yyyy-MM-dd HH:mm:ss,SSS
+    on_failure:
+    - append:
+        field: error.message
+        value: '{{ _ingest.on_failure_message }}'
+- date:
+    if: ctx?.zookeeper?.audit?.timestamp != null && ctx.event.timezone != null
+    field: zookeeper.audit.timestamp
+    target_field: '@timestamp'
+    formats:
+    - yyyy-MM-dd HH:mm:ss,SSS
+    timezone: '{{ event.timezone }}'
+    on_failure:
+    - append:
+        field: error.message
+        value: '{{ _ingest.on_failure_message }}'
+- kv:
+    field: message
+    field_split: "\\s+"
+    value_split: "="
+    target_field: zookeeper.audit
+- remove:
+    field:
+      - message
+      - zookeeper.audit.timestamp
+    ignore_missing: true
+- rename:
+    field: zookeeper.audit.operation
+    target_field: event.action
+    ignore_missing: true
+- set:
+    field: event.outcome
+    value: "{{zookeeper.audit.result}}"
+    ignore_empty_value: true
+    if: '["success","failure"].contains(ctx.zookeeper?.audit?.result)'
+- rename:
+    field: zookeeper.audit.ip
+    target_field: client.address
+    ignore_missing: true
+- convert:
+    field: client.address
+    target_field: client.ip
+    type: ip
+    ignore_missing: true
+- geoip:
+    field: client.ip
+    target_field: client.geo
+    if: ctx?.client?.ip != null
+- geoip:
+    database_file: GeoLite2-ASN.mmdb
+    field: client.ip
+    target_field: client.as
+    properties:
+    - asn
+    - organization_name
+    ignore_missing: true
+    if: ctx?.client?.ip != null
+# - split:
+#     field: zookeeper.audit.user
+#     separator: ","
+#     ignore_missing: true
```

Comment:
I guess this can be removed?

Reply:
Yes if we want to just leave the `user` field as a string.

Reply:
pushed

Reply:
Oh sorry, after your comment here https://github.com/elastic/beats/pull/25128#discussion_r618361589 I agree that it would make sense to have multiple users in the same event. So if this is the code that makes it, please recover it.

Reply:
so do we want to keep the IP as a "user"? I agree that its weird but again, I don't have any real logs using it, just the example logs.
```
        "user.name": [
            "192.168.1.2",
            "zkcli@HADOOP.COM"
        ]
```
or
```
        "user.name": "192.168.1.2,zkcli@HADOOP.COM",
```

Reply:
Yep, this is weird, but it seems to be the way to go with Zookeeper. I am thinking now that maybe we can keep both things:

* Store it as is, without splitting, in the `user.id` field. `user.id` is used for "unique identifiers", I guess that from the Zookeeper perspective, `192.168.1.2,zkcli@HADOOP.COM` and `172.24.0.2,zkcli@HADOOP.COM` refer to two different users, even if it is the same one connecting from two different IPs.
* Keep this also as a list, splitting the `user` field by commas as you were doing, but as this seems to be something quite Zookeeper-specific, don't use an ECS field for this. Use some field such as `zookeeper.audit.user`. Then this field can also be used to do queries per id at the authentication provider level.

In the future this can be revisited and do some additional processing to move or copy some of the users in `zookeeper.audit.user` to ECS fields. For example if any of these users is a system user, it could go to `user.name`, or if the user comes from the `X509AuthenticationProvider`, maybe it can be stored under `x509.subject.common_name`.

Wdyt?

Reply:
If you don't like the idea of keeping the same data basically twice, I would be fine with using one of the previous options: store the list in a custom field, or store the unsplitted field under `user.id`. And we could still revisit this in the future if needed.

Reply:
I think its a valid way ahead.  Updated code.